myapp/src/pages/LoginPage.js

```diff
@@ -0,0 +1,86 @@
+import { useState } from "react";
+import { useNavigate } from "react-router-dom";
+import BeatLoader from "react-spinners/ClipLoader";
+
+import InputField from "../components/InputField";
+
+function LoginPage() {
+  const [username, setUsername] = useState("");
+  const [password, setPassword] = useState("");
+
+  const [usernameError, setUsernameError] = useState("");
+  const [passwordError, setPassswordError] = useState("");
+
+  let [loading, setLoading] = useState(false);
+
+  const navigate = useNavigate();
+
+  function setData(event) {
+    event.preventDefault();
+
+    setUsernameError("");
+    setPassswordError("");
+
+    if (username === "") {
+      setUsernameError("Username is required!");
+      return;
+    } else if (password === "") {
+      setPassswordError("Password is required!");
+      return;
+    }
+
+    setLoading(!loading);
```

Comment:
This is not the best use of this flag. The intention is correct. We need to show a spinner when the form is being submitted. But since we are always toggling the flag instead of setting `true` or `false`, whenever this function runs it will first set to `true` and the next time it will be `false`. So for example if the function runs into an error and exists after toggling the flag, the spinner will continue to spin and as soon as we try to login again the spinner will dissapear (which is the reverse behaviour).

What we should do instead is set the flag to `true` at the start of the function, handle any errors within function with try catch statements, and then use a finally block to set the flag to `false`. This way we will start to spin whenever we click on submit, and the spinner will stop whenever the submit execution stops.

Reply:
Added try, catch and finally blocks and fixed it.
aa439fed25ecd2380580da968e5ee9fca0a39698